src/main/java/com/worksap/fig/lang/Seq.java

```diff
@@ -627,14 +653,133 @@ default boolean containsAny(Predicate<T> condition) {
+     * in this list, or -1 if this list does not contain the element.
+     * More formally, returns the lowest index <tt>i</tt> such that
+     * <tt>(o==null&nbsp;?&nbsp;get(i)==null&nbsp;:&nbsp;o.equals(get(i)))</tt>,
+     * or -1 if there is no such index.
+     */
+    int indexOf(T t);
+
+    /**
+     * Returns the index of the last occurrence of the specified element
+     * in this list, or -1 if this list does not contain the element.
+     * More formally, returns the highest index <tt>i</tt> such that
+     * <tt>(o==null&nbsp;?&nbsp;get(i)==null&nbsp;:&nbsp;o.equals(get(i)))</tt>,
+     * or -1 if there is no such index.
+     */
+    int lastIndexOf(T t);
+
+    /**
+     * Check whether this seq contains the sub seq, if the given seq is empty, always return true.
+     *
+     * @throws NullPointerException if seq is null
+     */
+    default boolean containsSubSeq(Seq<T> seq) {
+        return indexOfSubSeq(seq) != -1;
+    }
+
+    /**
+     * Return the first index of the given sub seq, or -1 if the given seq is not a sub seq.
+     * If the given seq is empty, always return 0.
+     *
+     * @throws NullPointerException if seq is null
+     */
+    default int indexOfSubSeq(Seq<T> seq) {
+        Objects.requireNonNull(seq);
+
+        if (seq.isEmpty()) {
+            return 0;
+        }
+
+        if (size() < seq.size()) {
+            return -1;
+        }
+
+        //Sunday algorithm
+
+        Map<T, Integer> lastIndex = new HashMap<>();
+        seq.forEach(lastIndex::put);
+
+        int startI = 0, size = size(), len = seq.size();
+        while (size - startI >= len) {
+            for (int i = 0; i < len; i++) {
+                if (!get(startI + i).equals(seq.get(i))) {
+                    if (startI + len >= size) {
+                        return -1;
+                    }
+                    T next = get(startI + len);
+                    Integer last = lastIndex.get(next);
+                    if (last == null) {
+                        startI += len + 1;
+                    } else {
+                        startI += len - last;
+                    }
+                    break;
+                } else if (i == len - 1) {
+                    return startI;
+                }
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * Return the last index of the given sub seq, or -1 if the given seq is not a sub seq.
+     * If the given seq is empty, always return 0.
+     *
+     * @throws NullPointerException if seq is null
+     */
+    default int lastIndexOfSubSeq(Seq<T> seq) {
+        Objects.requireNonNull(seq);
+
+        if (seq.isEmpty()) {
+            return 0;
+        }
+
+        if (size() < seq.size()) {
+            return -1;
+        }
+
+        //Sunday algorithm
+
+        Map<T, Integer> lastIndex = new HashMap<>();
+        seq.forEachReverse(lastIndex::put);
+
+        int size = size(), endI = size - 1, len = seq.size();
```

Comment:
size is no used latter. Size and len are easy to be confused. How about thisSize and ptnSize?